### PR TITLE
Fix async_subscribe lifetime management

### PR DESCRIPTION
**Pull request by jonesmz**

### CODE REVIEW

**Comment by jonesmz** [comment]
I had to add several updates to this.

It turns out we weren't properly managing the lifetime of the async_subscribe functions that took as::const_buffer. So I added the proper life_keeper parameters, and updated the unit tests.

**Comment by redboltz** [comment]
These async functions has async_handler_t, and user can capture life keeping object to the lambda expression. That is the original design. And test code did that.
But it seems to be difficult to understand. In addition, the user set not only the lambda expression but also function pointer as the async_handler_t. So I think that your update is better.

Could you remove the original capture code?

File: test/as_buffer_async_pubsub_1.cpp [hunk 1]
```diff
@@ -47,6 +47,7 @@ BOOST_AUTO_TEST_CASE( pub_qos0_sub_qos0 ) {
                     pid_sub = c->async_subscribe(
                         as::buffer(*topic1),
                         MQTT_NS::qos::at_most_once,
+                        topic1,
                         [topic1](boost::system::error_code const&) {}
```
Comment (redboltz) [inline test/as_buffer_async_pubsub_1.cpp]: This is an example of the capture code.

**Comment by jonesmz** [comment]
The standard way for doing lifetime management with boost.asio is to provide a completion token, not a std::function.

My original reason for suggesting that we add the `life_keeper` parameter to the functions that need to preserve parameter lifetime was two-fold.

1. Internally, mqtt_cpp adds additional stuff to the async_handler_t function by wrapping the function in (potentially) several layers of lambdas. That's very awkward from an internal implementation perspective. Making the incoming data types more generic (e.g. a pair of iterators, I think is best. Most flexible), so that `std::vector<MQTT_NS::buffer>` is e.g. `boost::iterator_range<TYPE_T>` that would allow converting from type `TYPE_T` -> `boost::asio::buffer` without any allocations, or nesting of lambda functions. This PR, and https://github.com/redboltz/mqtt_cpp/pull/524 help with this.
2. I don't want to use lambda functions as the completion token. I'd rather use something like Boost.Coroutine, but since mqtt_cpp only takes std::function as the callback, I can't currently. The `any life_keeper` parameter is easier for my code, but if I could use Boost.Coroutine, I would rather do the lifetime with the CompletionToken. See https://github.com/redboltz/mqtt_cpp/issues/523

When 1) and 2) are solved, we can remove the `any life_keeper` parameter.

File: test/as_buffer_async_pubsub_1.cpp (hunk 1 above)
Comment (jonesmz) [reply test/as_buffer_async_pubsub_1.cpp]: Updated to remove lambda-capture variables.

**Comment by jonesmz** [comment]
Following up on my previous comment, I guess we need to make a decision.

1) Use the "life_keeper" way of preserving the lifetime of the data for the message for everything that needs it.
2) Do not use "life_keeper", so back to boost.asio method where lifetime is bundled with async_handler_t / CompletionToken.

If we can do: https://github.com/redboltz/mqtt_cpp/issues/523, and https://github.com/redboltz/mqtt_cpp/issues/536

Then we do not need the idea of "life_keeper", and can remove it, and use only async_handler_t / CompletionToken.

**Comment by redboltz** [comment]
I'm on some trouble with my project. So I don't have much time to consider about that.

But I just roughly read #523, #536, and #534 .

1. Basically, I think that they are nice idea. Let's go on that way.
2. I don't want to change frequently end-user APIs even if it is master branch. If those update can be done all at once, then end-user updates their code only once. But I think that the updates are too large to one PR, it is difficult to review.  So I create the branch for the work. Maybe the branch name is `generic_api`. You can send PRs for `generic_api` that solves #523, #536, and #534 individually. And this PR (#532) is also for `generic_api` branch.
3. When developing `generic_api` is finished, then merge `generic_api` to master.

I'd like to use this process. What do you think?

**Comment by jonesmz** [comment]
> I'd like to use this process. What do you think?

All at once is fine with me.

We still need to submit part of this pull request to resolve a compiler error that I'm experiencing though.

Let me update the PR to only fix the compile problem.

**Comment by redboltz** [comment]
> All at once is fine with me.

Thanks. It can avoid frequently API updates on master.

Let me clarify one thing.
Do you want to add commits for #523, #536, and #534 fix to the PR (#532) ?
I can create working branch. Don't you need it?

> We still need to submit part of this pull request to resolve a compiler error that I'm experiencing though.
> 
> Let me update the PR to only fix the compile problem.

Ok.

**Comment by jonesmz** [comment]
> I can create working branch. Don't you need it?

I will work on my repository, and submit PR all at once.

**Comment by jonesmz** [comment]
@redboltz Is it possible to merge this change soon? It fixes a compiler error that I'm encountering when trying to use the master branch.

**Comment by jonesmz** [comment]
I removed many changes from this PR, because we discussed doing:

>  #523, #536, and #534 fix all at once?

This PR is not only for fixing compiler error

**Comment by redboltz** [comment]
> I removed many changes from this PR, because we discussed doing:
>
> > #523, #536, and #534 fix all at once?
>
> This PR is now only for fixing compiler error

I understand. I will merge it soon.
I misunderstood that you would add #523, #536, and #534 fixes to #532.

**Pull request merged**